libopenage/renderer/font_manager.h

```diff
@@ -0,0 +1,86 @@
+// Copyright 2015-2015 the openage authors. See copying.md for legal info.
+
+#ifndef OPENAGE_RENDERER_FONT_MANAGER_H_
+#define OPENAGE_RENDERER_FONT_MANAGER_H_
+
+#include <ft2build.h>
+#include FT_FREETYPE_H
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+#include "../error/error.h"
+
+namespace openage {
+namespace renderer {
+
+struct freetype_library {
+	FT_Library ft_library;
+
+	freetype_library() {
+		if (FT_Init_FreeType(&this->ft_library) != 0) {
+			throw Error{MSG(err) << "Failed to initialize freetype library."};
+		}
+	}
+
+	~freetype_library() {
+		FT_Done_FreeType(this->ft_library);
+	}
+};
+
+class Font;
+
+class FontManager {
+public:
+	/**
+	 * Get the freetype library instance.
+	 */
+	static FT_Library get_ft_library();
+
+	/**
+	 * Gets the filepath of a particular font family and style.
+	 *
+	 * @param family: The font family.
+	 * @param style: The font style.
+	 * @returns The path to font's file.
+	 */
+	static std::string get_font_filename(const char *family, const char *style);
+
+public:
+	FontManager();
+
+	~FontManager();
+
+	/**
+	 * Retrieves a font.
+	 *
+	 * @param family: The font family.
+	 * @param style: The font style.
+	 * @param size: The size of the font in points.
+	 * @returns The pointer to font instance.
+	 */
+	Font *get_font(const char *family, const char *style, unsigned int size);
+
+	/**
+	 * Retrieves a font.
+	 *
+	 * @param font_file: The path to font's file.
+	 * @param size: The size of the font in points.
+	 * @returns The pointer to font instance.
+	 */
+	Font *get_font(const char* font_file, unsigned int size);
+
+private:
+	// Font cache. the hash of font's description is used as the key
+	std::unordered_map<size_t, std::unique_ptr<Font>> fonts;
+
+private:
+	// The freetype library instance
+	static freetype_library library;
```

Comment:
FreeType provides the option to create multiple instance of `FT_Library`. Typically, you would one library instance to manage all the font faces loaded in this library.

The API reference also adds additional notes with respect to multithreading
> A handle to a FreeType library instance. Each ‘library’ is completely independent from the others; it is the ‘root’ of a set of objects like fonts, faces, sizes, etc.

> In multi-threaded applications it is easiest to use one ‘FT_Library’ object per thread. In case this is too cumbersome, a single ‘FT_Library’ object across threads is possible also (since FreeType version 2.5.6), as long as a mutex lock is used around FT_New_Face and FT_Done_Face.